### PR TITLE
engineering: Blue-green Deployment

### PR DESCRIPTION
**Pull request by nguyennh4522**
#### What's this PR do?

- [x] Explain what is Blue-green Deployment
- [x] Explain how to implement blue-green deployment
- [x] Compare with other strategies

### CODE REVIEW

**Review by lmquang — APPROVED** [review]

**Comment by lmquang** [comment]
@monotykamary  xem qua r merge giup a nha e

File: Engineering/DevOps/Blue-green Deployment.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
+---
+tags: engineering/devops
+author: Nguyen Huu Nguyen
+github_id: nguyennh4522
+date: 2022-02-16
+---
+
+**Blue-green** deployment is a software deployment strategy that involves creating two identical environments: one **blue** environment serving production traffic, and another **green** environment that doesn't serve any traffic. Once the **green** environment is fully tested and verified, traffic is switched from the **blue** environment to the **green** environment, making it the new production environment. This approach reduces downtime, improves reliability and resilience, and provides a backup in case of issues.
+
+![[blue-green-deployment-model.gif]]
```
Comment (toanbku) [inline Engineering/DevOps/Blue-green Deployment.md:10]: should be `alt of image`

File: Engineering/DevOps/Blue-green Deployment.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+## How does blue-green deployment work?
+
+**Blue-green** deployment is a deployment strategy for software applications that involves maintaining two identical environments: one currently serving production traffic (the **blue** environment), and one that is newly deployed (the **green** environment"). The new version of the application is deployed to the **green** environment, which is tested and monitored. Once it is determined that the **green** environment is working correctly, traffic is routed to it and the **blue** environment is retired. This strategy allows for quick and easy switching between environments, minimizing downtime and reducing the risk of errors or bugs.
+
+## Characteristic
+
+The following table summarizes the salient features of the **blue-green** strategy compared to other strategies:
+
+![[bluegreen-compare.png]]
```
Comment (toanbku) [inline Engineering/DevOps/Blue-green Deployment.md:20]: same here

File: Engineering/DevOps/Blue-green Deployment.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+          image: argoproj/rollouts-demo:blue
+...
+```
+```bash
+kubectl apply -f app/bluegreen-rollout.yaml
+```
+
+Argo rollouts will create a new ReplicaSet with the new image and start to rollout the new version of the application to the preview environment. You can view the preview application on `myapp-preview.bluegreen.xyz` domain.
+
+![[bluegreen-blue-application.png]]
```
Comment (toanbku) [inline Engineering/DevOps/Blue-green Deployment.md:198]: same here

File: Engineering/DevOps/Blue-green Deployment.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+```bash
+kubectl apply -f app/bluegreen-rollout.yaml
+kubectl apply -f app/service.yaml
+kubectl apply -f app/ingress.yaml
+```
+
+Now, application is deployed on active environment and you can view it on `myapp.bluegreen.xyz` domain.
+
+![[bluegreen-green-application.png]]
```
Comment (toanbku) [inline Engineering/DevOps/Blue-green Deployment.md:180]: same here

**Review by toanbku — CHANGES_REQUESTED** [review]

**Review by zlatanpham — APPROVED** [review]

**Pull request merged**